### ISSUE TITLE
Request: list of topics, list/number of partitions and watermark offsets for each 

### ISSUE THREAD

**Issue by martindurant** [confluentinc/confluent-kafka-python#330]
Description
===========

Would love to see the following functions added:

- list topics on the server
- for a given topic, fetch the number of partitions
- for a given topic, fetch the watermark offsets for *all* partitions in a single call (rather than looping over each).

Please advise if these are already implemented (in which case, sorry!) or whether they are at least possible in the current framework.

**Comment by mrocklin** [comment]
I would also find these valuable.  When I looked last time all of these were possible to do, but took some time.  In particular I think that the mechanism to query offset was restricted to a single partition, and so I found myself looping over partitions, and each call took longer than expected.  

My understanding was that all of this information comes in a single message from kafka, so my hope is that it's at least feasible, even if not possible today from the Python API.

**Comment by martindurant** [comment]
https://cwiki.apache.org/confluence/display/KAFKA/A+Guide+To+The+Kafka+Protocol#AGuideToTheKafkaProtocol-OffsetRequest suggests that the partition is always required to get the range of offsets (presumably because each partition may be led by a different broker).

**Comment by edenhill** [comment]
There's a PR #309 by @tbsaunde to add the list_topics() API that will give you 1 and 2 in your list.
As for get_watermarks, yes, it will need to be improved to support multiple partitions in the same call (in librdkafka and Python)

**Comment by edenhill** [comment]
You'll need to convert the partition list to a `[TopicPartition()..]` list prior to calling position().
Also note that position() returns the current position for an active assignment, not the committed offsets (committed() does that)

**Comment by jacopofar** [comment]
Right, now to the the initial offset of a partition I use `client.get_watermark_offsets(TopicPartition(topic, p_index))[0]` where topic is the topic name and p_index the numeric index of the partition.

**Comment by martindurant** [comment]
Right, that works, but bringing this back to the initial request, this all requires several calls, whereas it seems it should be possible to get all partition offsets in a single call.

**Comment by edenhill** [comment]
In a future release we will improve get_watermark_offsets() to take a list of partitions instead of a single one.

**Comment by martindurant** [comment]
Was this implemented?

**Comment by edenhill** [comment]
No, closed this due to inactivity/not that much interest from community.
Reopen if still needed.

**Comment by martindurant** [comment]
I cannot reopen it. Yes, it is an essential feature for low-latency polling.

**Comment by martindurant** [comment]
@chinmaychandak @jsmaupin perhaps you have some benchmarking to prove the point.

**Comment by chinmaychandak** [comment]
Yes, it would be really great if we could support a list of partitions in `get_watermark_offsets`. 

In case of the streaming use case Martin mentioned, we poll Kafka for a topic's metadata as well as for the watermark offsets for all the topic's partitions a few times every second. 

Upon doing some profiling, I noticed that a single `get_watermark_offsets` call was taking about `~2.5ms` on AWS with Kafka and the consumer being within the same VPC but on different instances. This is much larger than the latency incurred in retrieving a message, listing topics, etc - all `<1ms` individually. And since loop to retrieve watermark offsets for all partitions in every polling cycle (few cycles/sec as mentioned above), it adds on as a substantial overhead.

**Comment by edenhill** [comment]
If you want to monitor the high watermark of the partitions a consumer is assigned to you can use get cached version, which does not require a roundtrip to the broker.

**Comment by chinmaychandak** [comment]
If I understand correctly, the cached version ensures periodic updates from the broker - but since we need the latest high watermark for real-time data consumption/streaming, we set cached to False.

**Comment by edenhill** [comment]
On each FetchRequest to the broker it will cache the high watermark, so this should be as up to date as it gets for an active consumer.
The Low watermark is queried every max(10s, statistics.interval.ms) (it will change much less frequently than that)

**Comment by chinmaychandak** [comment]
I see, in that case I will try out using cached to True, and benchmark again. Thank you so much!

Although, would it be too much work in adding the support to a list of partitions in `get_watermark_offsets`? It would be great to have the same support in this API as `commit`, `committed`, etc. Not having to loop to get watermark offsets for multiple partitions would be really valuable.
